gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/medication/MedicationRequestMapper.java

```diff
@@ -132,6 +143,50 @@ private static void updateBasedOnReferenceToReferenceDuplicatedPlan(List<Referen
         }
     }
 
+    private @NotNull MedicationStatement duplicateMedicationStatementForOrderBasedOnSharedAcutePlan(
+        List<MedicationStatement> medicationStatements,
+        MedicationRequest order,
+        MedicationRequest originalPlan,
+        MedicationRequest duplicatedPlan
+    ) {
+
+        var originalMedicationStatement = medicationStatements
+            .stream()
+            .filter(medicationStatement -> medicationStatement.getId().equals(originalPlan.getId() + "-MS"))
+            .findFirst()
+            .orElseThrow(() ->
+                // there should always be a medication statement associated with an order, but we have to handle
+                // the possibility it is not found.
+                new IllegalStateException(
+                    "MedicationStatement referenced from MedicationRequest[Order] (%s) is not found."
+                        .formatted(order.getId())
+                )
+            );
+
+        var duplicatedMedicationStatement = originalMedicationStatement.copy();
+
+        // we need to update the id and identifier to the generated id;
+        duplicatedMedicationStatement.setId(duplicatedPlan.getId() + "-MS");
+        duplicatedMedicationStatement.getIdentifierFirstRep().setValue(duplicatedPlan.getId() + "-MS");
+
+        // we need to update the basedOn of the order to match the generated plan
+        updateBasedOnReferenceToReferenceDuplicatedPlan(
+            duplicatedMedicationStatement.getBasedOn(),
+            duplicatedPlan.getId()
+        );
+
+        // we need to update the effective period to match that of dispenseRequest validityPeriod from the order.
+        duplicatedMedicationStatement.setEffective(order.getDispenseRequest().getValidityPeriod());
+
+        // finally, we need to update the LastIssueDateExtension to match the dispenseRequest validity period start from
+        // the order.
+        duplicatedMedicationStatement
+            .getExtensionByUrl(MEDICATION_STATEMENT_LAST_ISSUE_DATE_URL)
+            .setValue(order.getDispenseRequest().getValidityPeriod().getStartElement());
+
+        return duplicatedMedicationStatement;
+    }
```

Comment:
Another large method, in the spirit of keeping methods small - is there any way you can see fit to break this up into smaller chunks - that of course would make sense to do so?

Reply:
Removed the unneeded comments and moved the medication statement find to a new method, but the rest is all the fields that need to be updated during the duplication, so it makes sense to keep them in here